### PR TITLE
Fix regression around CSS stacking contexts and PIP widgets

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/element-hq/element-web/issues/26805

Regressed by https://github.com/matrix-org/matrix-react-sdk/pull/12052

It added a rule `#matrixchat { contain: strict; }` which changed how the layout engine works.
It enabled CSS stacking contexts, which z-index cannot cross, its like a 2nd order z-index, the buttons were in the wrong stacking context for where they wanted to be drawn, so the PR moves them to the correct context (by moving it into the PersistentElement as that's a React Portal which renders outside of #matrixchat)

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix regression around CSS stacking contexts and PIP widgets ([\#12094](https://github.com/matrix-org/matrix-react-sdk/pull/12094)). Fixes element-hq/element-web#26805.

### CODE REVIEW

**Comment by toger5** [comment]
Thank you so much!
There are issues with the styling:
 - The border radius of 8px is missing: this also missing in the css right now but will be added back in https://github.com/matrix-org/matrix-react-sdk/pull/12057/files
 This also is not working witht he new drawing anymore (I tried adding the radius in the devtools but the IFrame did not accept :`border-radius: 8px;overflow: hidden;` A possible workaround would be to add a 0.989 opacity to the IFrame.
 - The hangup and back button and footer/header gradients are too dark

**Comment by t3chguy** [comment]
> The hangup and back button and footer/header gradients are too dark

I cannot reproduce this.

Before
image

After
image

Can reproduce the border radius issue though so will fix that.

After^2
image

**Comment by toger5** [comment]
This is what it looks for me:
Screenshot 2024-01-02 at 12 05 41
On firefox. Looking through the css it seems like there is no explicit color css property on the hangup icon and the title text.

**Comment by t3chguy** [comment]
@toger5 cannot reproduce.

Before
image

After
image

Actually comparing the screenshots looks like app.element.io is darker?

**Comment by toger5** [comment]
> This is what it looks for me: Screenshot 2024-01-02 at 12 05 41 On firefox. Looking through the css it seems like there is no explicit color css property on the hangup icon and the title text.

I also tried on chrome where it looks the same as on FF.
Are you trying with you local build or with netlify? I am on trying with the netlify build. Maybe some dependencies are different.

**Comment by fkwp** [comment]
image

same issue for me on safari using the netlify build

**Comment by toger5** [comment]
> If this was a regression, could you link the PR that caused it, to help understand the fix? Or generally add some detail to the issue about how the problem came to be?

The problem came to be because of using the strict contained mode as described in the issue: https://github.com/element-hq/element-web/issues/26805#issue-2054362601.
The regression is: no interaction and wrong styling with pip elements.

@fkwp and me are having issues with the fix however that there are styling issues: (icon and text color is too dark)

**Comment by richvdh** [comment]
> The problem occured because of using the strict contained mode as described in the issue: https://github.com/element-hq/element-web/issues/26805#issue-2054362601.

I'm afraid I still don't really follow what changed.

> @fkwp and me are having issues with the fix however that there are styling issues: (icon and text color is too dark). But @t3chguy cannot reproduce those.

Sounds like this is still in development, so moving this to draft for now.

**Comment by t3chguy** [comment]
image

Looks like its specific to light theme.

image

Moved the colour to fix.

**Review by toger5 — APPROVED** [review]
Great! Thanks for fixing the light theme.

File: res/css/_common.pcss [hunk 1]
```diff
@@ -56,8 +56,10 @@ limitations under the License.
     /* This is required to ensure Compound tooltips correctly draw where they should with z-index: auto */
     contain: strict;
 }
-.mx_Dialog_StaticContainer,
-.mx_Dialog_Container {
+#mx_ContextualMenu_Container,
+#mx_PersistedElement_container,
+#mx_Dialog_Container,
+#mx_Dialog_StaticContainer {
```
Comment (richvdh) [inline res/css/_common.pcss:62]: why are these changing?

File: src/components/views/elements/AppTile.tsx [hunk 2]
```diff
@@ -93,6 +93,8 @@ interface IProps {
     showLayoutButtons?: boolean;
     // Handle to manually notify the PersistedElement that it needs to move
     movePersistedElement?: MutableRefObject<(() => void) | undefined>;
+    // An element to render after the iframe as an overlay
+    overlay?: ReactNode;
```
Comment (richvdh) [inline src/components/views/elements/AppTile.tsx:97]: wouldn't it make more sense for this to be passed as `children`, rather than a regular prop?

File: res/css/views/rooms/_AppsDrawer.pcss [hunk 3]
```diff
@@ -325,6 +325,9 @@ limitations under the License.
     &.mx_AppTileBody--call {
         border-radius: 0px;
     }
+    &.mx_AppTileBody--call.mx_AppTileBody--mini {
+        border-radius: 8px;
+    }
```
Comment (richvdh) [inline res/css/views/rooms/_AppsDrawer.pcss:330]: why is this now needed (as well as the border-radius on `.mx_WidgetPip_overlay ` ?

(I have no real idea what a `&.mx_AppTileBody--call.mx_AppTileBody--mini` is, I'm afraid)

File: src/components/structures/PipContainer.tsx [hunk 4]
```diff
@@ -305,6 +305,7 @@ class PipContainerInner extends React.Component<IProps, IState> {
             const call = this.state.primaryCall;
             pipContent.push(({ onStartMoving, onResize }) => (
                 <LegacyCallView
+                    key="call-view"
```
Comment (richvdh) [inline src/components/structures/PipContainer.tsx:308]: any particular reason these `key`s are now needed?

File: test/setupTests.ts [hunk 5]
```diff
@@ -38,9 +38,11 @@ beforeEach(() => {
 });
 
 // Retry to work around our flaky app & tests
-jest.retryTimes(2, {
-    logErrorsBeforeRetry: true,
-});
+if (process.env.CI) {
```
Comment (richvdh) [inline test/setupTests.ts:41]: I'm thinking this should be in a separate PR

**Review by richvdh — COMMENTED** [review]
I have no real idea what's going on here, I'm afraid. It looks vaguely plausible. Since I don't understand the nuclear reactor, I have some questions about the colour you have decided to paint the bikeshed.

File: res/css/_common.pcss (hunk 1 above)
Comment (t3chguy) [reply res/css/_common.pcss:62]: So that they layer correctly relative to the PIP widget and tooltips and each other by having their own CSS stacking contexts

File: src/components/views/elements/AppTile.tsx (hunk 2 above)
Comment (t3chguy) [reply src/components/views/elements/AppTile.tsx:97]: Children have different semantics of typically always being included, this only includes it when the iframe is shown, not when the loader or permissions prompt are shown

File: src/components/structures/PipContainer.tsx (hunk 4 above)
Comment (t3chguy) [reply src/components/structures/PipContainer.tsx:308]: They fix React warnings from this area of code, they were always needed

https://github.com/matrix-org/matrix-react-sdk/pull/12094/commits/977216cb90ddcde4098616570047d9f31798267e

File: test/setupTests.ts (hunk 5 above)
Comment (t3chguy) [reply test/setupTests.ts:41]: Sure

File: res/css/views/rooms/_AppsDrawer.pcss (hunk 3 above)
Comment (toger5) [reply res/css/views/rooms/_AppsDrawer.pcss:330]: Before the widget was always rendered the default app container with rounded borders.
This is a concious design decision to underline that it is not part of Eement web but potentially a thrid party app running as a widget. For calls we want to special case this and show the widget as if its part of Element Web.
So `mx_AppTileBody--call` has `border-radius: 0px;`.
For the PiP call view rounded corners are alwasy required. So we set it back to 8px.

Old:
Screenshot 2024-01-05 at 21 49 58
Now:
Screenshot 2024-01-05 at 21 56 58
Issue caused by new design
Screenshot 2024-01-05 at 21 53 16
Fix for new design:
Screenshot 2024-01-05 at 21 53 49

File: test/setupTests.ts (hunk 5 above)
Comment (t3chguy) [reply test/setupTests.ts]: [Done](https://github.com/matrix-org/matrix-react-sdk/pull/12111)

File: res/css/views/rooms/_AppsDrawer.pcss (hunk 3 above)
Comment (richvdh) [reply res/css/views/rooms/_AppsDrawer.pcss:330]: > Before the widget was always rendered the default app container with rounded borders.

I don't follow. There don't seem to be enough words in this sentence?

> alwasy

always

File: src/components/views/elements/AppTile.tsx (hunk 2 above)
Comment (richvdh) [reply src/components/views/elements/AppTile.tsx:97]: > Children have different semantics of typically always being included

I'm not entirely convinced about that. And if it's true, doesn't the same logic apply to `PersistentApp`?

But I don't feel strongly about it.

**Review by richvdh — APPROVED** [review]
seems plausible to me
